classloader-leak-prevention/classloader-leak-prevention-core/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/ThreadLocalCleanUp.java

```diff
@@ -94,7 +94,9 @@ protected void processThreadLocalMap(ClassLoaderLeakPreventor preventor,
             java_lang_ThreadLocal$ThreadLocalMap$Entry_value = preventor.findField(entry.getClass(), "value");
           }
 
-          final Object value = java_lang_ThreadLocal$ThreadLocalMap$Entry_value.get(entry);
+          // Dereference the value if this is a Reference<T>: all Reference<T> implementations are all loaded using the bootstrap classloader,
+          // so checking the Reference<T> classloader won't indicate if the held value was itself loaded using the app classloader 
+          final Object value = dereferenceIfApplicable(java_lang_ThreadLocal$ThreadLocalMap$Entry_value.get(entry));
```

Comment:
Would it have made more sense to `.clear()` the `Reference` in a case like this?

Reply:
Not sure what you mean - it looks the value is retrieved only to check whether it's related to the classloader and trigger the cleanup in `se.jiderhamn.classloader.leak.prevention.cleanup.ThreadLocalCleanUp.processLeak(ClassLoaderLeakPreventor, Thread, Reference<?>, ThreadLocal<?>, Object, String)` - which actually does not make use of the `value` at all

Did I miss something?

Reply:
What I mean here, if `java_lang_ThreadLocal$ThreadLocalMap$Entry_value.get(entry)` is `instanceof Reference`, then maybe we want to dereference (recursively) and if the dereferenced value is `preventor.isLoadedInClassLoader(dereferencedValue)`, then we may want to call `.clear()` on the `Reference` __instead of__ setting `valueLoadedInWebApp` to true and calling `processLeak()`.

That would be a less intrusive way to prevent the leak. What do you think?

Reply:
Ah ok, got it now.
Yes, this is indeed possible as well.
However, this would leave a dangling `ThreadLocal` with a cleared `Reference<?>`.
I'd rather clear the `ThreadLocal` invoking `processLeak` (like done now in the current state of this PR) to avoid leaving this dangling entry behind, and this also has the advantage to keep the code clean (meaning without an additional edge case) and more maintainable.
Don't you think?

Reply:
The dangling entry will be left behind either way, in order not to interfere with how the array of `ThreadLocal`s are stored in the `Thread`. What I'm thinking is that a `Reference.clear()` should be somewhat more likely to succeed, while the value clear on line 201 could potentially fail with `IllegalAccessException`.

I don't think we should `clear()` instead, but possibly we could do it _also_.

Suggestion: For now, just make a comment that we could have `clear()`ed it here. Sounds ok?